File: langton_ant/langton_ant.cc
#include <stdio.h>
#include <thread>
#include <chrono>

const int width = 20;
const int height = width;

const int max_iterations = 10'000;
// up to 0 - 1000 iterations per second
// you can increase iterations to speed up the simulation and see more patterns
// but increasing it too much makes rendering look bad because the ansi functions can't keep up
const int iterations_per_second = 3;
int board[width*height]{1};

enum direction {
    left,
    up,
    right,
    down
};
// start in center
int ant_x = width / 2;
int ant_y = height / 2;
int ant_dir = up;
int current_move = 0;    

void turn_right(int& dir) {
    dir = dir == down ? left : dir+1;
}

void turn_left(int& dir) {
    dir = dir == left ? down : dir-1;
}

void move(int& ant_x, int& ant_y, int ant_dir) {
    switch ( ant_dir ) {
        case left: ant_x--;break;
        case up: ant_y--;break;
        case right: ant_x++;break;
        case down: ant_y++;break;
    }
}

void wrap(int& n) {
    if ( n >= width )
        n = 0;
    else if (n <= 0 )
        n = width - 1;
}

enum color {
    white,
    black
};

void red(bool set) {
    if (set)printf("\x1b[31m");
    else printf("\x1b[0m");
}

void clear_screen() {
    printf("\x1b[1J");
}

void reset_cursor() {
    printf("\x1b[1;1H");
}
void draw_board() {
    clear_screen();
    reset_cursor();
    printf("\nmove: %d\n", ++current_move);
    for ( auto y = 0; y < height; y++ ) {
        for ( auto x = 0; x < width; x++ ) {
            // if this is the ant
            if ( y == ant_y and x == ant_x ) {
                red(true);
                switch ( ant_dir ) {
                    case left: putchar('<'); break;
                    case up: putchar('A'); break;
                    case right: putchar('>'); break;
                    case down: putchar('V'); break;
                }
                red(false);
            } else {
                if ( board[ y * width + x ] == white)
                    putchar('.');
                else
                    putchar('#');
            }
        }
        puts("");
    }
    // delay
    std::this_thread::sleep_for(std::chrono::milliseconds(1000/iterations_per_second));
}
int main() {
    draw_board();
    while ( true ) {
        auto pos = ant_y * width + ant_x;
        if ( board[pos] == white ) {
            turn_right(ant_dir);
            board[pos] = black;
            move(ant_x, ant_y, ant_dir);
        } else {
            turn_left(ant_dir);
            board[pos] = white;
            move(ant_x, ant_y, ant_dir);
        }

        wrap(ant_x);
        wrap(ant_y);
        draw_board();
    }

}

File: langton_ant/langton_ant.go
package main

import (
	"fmt"
	"time"
)
const (
	width = 20
	height = width
	maxIterations = 10_000
	// up to 0 - 1000 iterations per second
	// you can increase iterations to speed up the simulation and see more patterns
	// but increasing it too much makes rendering look bad because the ansi functions can't keep up
	iterationsPerSecond = 3
)

type Direction int
const (
	// directions
	left Direction = iota
	up
	right
	down
)

const (
	white = iota
	black
)

var (
	board [width * height]int
	// start in center
	antX = width / 2
	antY = height / 2
	antDir = up
	currentMove = 0
)

func main() {
	// init board to 0
	for _, i := range board {
		board[i]=white
	}

    drawBoard()
    for true  {
        pos := antY * width + antX
        if board[pos] == white {
            turnRight(&antDir)
            board[pos] = black
            move(&antX, &antY, antDir)
        } else {
            turnLeft(&antDir)
            board[pos] = white
            move(&antX, &antY, antDir)
        }

        wrap(&antX)
        wrap(&antY)
        drawBoard()
    }
}

func turnRight(dir *Direction) {
	if *dir == down {
		*dir = left
	} else {
		*dir += 1
	}
}

func turnLeft(dir *Direction) {
	if *dir == left {
		*dir = down
	} else {
		*dir -=1
	}
}

func move(antX, antY *int, antDir Direction) {
    switch antDir  {
        case left: 
			*antX -= 1
        case up: 
			*antY -= 1
        case right: 
			*antX += 1
        case down: 
			*antY += 1
    }
}

func wrap(n *int) {
    if  *n >= width  {
        *n = 0;
	} else if *n <= 0 {
        *n = width - 1;
	}
}

func red(set bool) {
	if set {
		fmt.Print("\033[31m")
	} else {
    	fmt.Print("\033[0m")
	}
}

func clearScreen() {
    fmt.Print("\x1b[1J")
}

func resetCursor() {
    fmt.Print("\x1b[1;1H")
}

func drawBoard() {
    clearScreen()
    resetCursor()
    fmt.Printf("\nmove: %d\n", currentMove)
	currentMove += 1
    for y := 0; y < height; y += 1 {
        for x := 0; x < width; x += 1 {
            // if this is the ant
            if y == antY && x == antX  {
                red(true)
                switch antDir {
                    case left: 
						fmt.Print("<")
                    case up: 
						fmt.Print("A")
                    case right: 
						fmt.Print(">")
                    case down: 
						fmt.Print("V")
                }
                red(false)
            } else {
                if board[ y * width + x ] == white {
                    fmt.Print(".")
				} else {
                    fmt.Print("#")

				}
            }
        }
		fmt.Println("")
    }
    // delay
	time.Sleep(time.Millisecond * (1000/iterationsPerSecond))
}